### PR TITLE
nheko: 0.6.4 -> 0.7.1

### PR DESCRIPTION
**Pull request by doronbehar**
~~DONT MERGE, this PR should work once https://github.com/NixOS/nixpkgs/pull/73940 is merged.~~

###### Motivation for this change

https://github.com/Nheko-Reborn/nheko/releases/tag/v0.7.1

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rnhmjoj** [comment]
Oh, it has been released, finally. I had prepared the update but you beat me to it.
I don't remember having problems with boost but I'm on 20.03, here's the [commit](https://maxwell.ydns.eu/git/rnhmjoj/nheko/commit/e3f2eb4d977dc75273ae2b9232bcfe63cd0cfbc1) maybe it can be of help.

**Comment by doronbehar** [comment]
One difference I spotted is the use of `boost171` vs `boost172` ( 17x is used here). I tested with boost171 and still the same error. @rnhmjoj what's the meaning of:

```
    "-DBoost_NO_BOOST_CMAKE=TRUE"
```

[here](https://maxwell.ydns.eu/git/rnhmjoj/nheko/src/commit/e3f2eb4d977dc75273ae2b9232bcfe63cd0cfbc1/mtxclient.nix#L19)?

It fixes the issue but it feels hacky to me considering https://github.com/NixOS/nixpkgs/pull/73940 ...

**Comment by veprbl** [comment]
@doronbehar Did you try #85254?

**Comment by doronbehar** [comment]
@veprbl https://github.com/NixOS/nixpkgs/pull/85254 works just as well as https://github.com/NixOS/nixpkgs/pull/73940

**Comment by bqv** [comment]
I've been using 0.7.0ish for some days, i don't remember having an issue with mtxclient building?

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 1]
```diff
@@ -48,23 +49,25 @@ mkDerivation rec {
...
+    mtxclient
+    olm
+    boost17x
+    lmdb
+    spdlog
+    cmark
+    qtbase
+    qtmultimedia
+    qttools
+    qtquickcontrols2
```
Comment (bqv) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:70]: Have you tested this? I think you'll have to add qtgraphicaleffects to the inputs too

Comment (doronbehar) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:70]: I did manage to make it launch and login but I had nothing showing up in conversations and I assumed it's an upstream bug. Now with `qtgraphicaleffects` it's added to the environment of the wrapper and the conversations show themselves so :+1:. Thanks.

**Comment by bqv** [comment]
Looks reasonable compared to the version I'm running, and nothing missing in that inputs list according to @deepbluev7

**Review by bqv — APPROVED** [review]

**Comment by veprbl** [comment]
@GrahamcOfBorg build nheko mtxclient tweeny

**Comment by bqv** [comment]
> I've been using 0.7.0ish for some days, i don't remember having an issue with mtxclient building?

Nevermind, I remember it now.

@doronbehar You may find this useful - I'm running and talking with this as we speak. Specifically I think the part that fixes your boost issue is the symlinkJoin
https://github.com/bqv/nixos/blob/live/overlays/nheko.nix

**Comment by doronbehar** [comment]
[That's no good](https://github.com/bqv/nixos/blob/b13457c1a4f1a6b76ac18a8d569c3f7261f9aa61/overlays/nheko.nix#L5) - I'm pretty sure that your final derivation references both `boost.dev` and `.out` - Not good for closure size.

**Comment by doronbehar** [comment]
@veprbl the build won't work until either #85254 or #73940 are merged. I don't want to add:

```
    "-DBoost_NO_BOOST_CMAKE=TRUE"
```

Which fixes the issue here because that's not a real fix. Please :pray: consider merging either #85254 or #73940.

**Comment by veprbl** [comment]
@doronbehar I can't just blindly merge any of these and I haven't had time yet to look at those in detail. I propose you use the workaround for now with a comment that it can be removed once #85254 and #73940 are resolved.

**Comment by deepbluev7** [comment]
Btw, you may want to include this commit in your package: https://github.com/Nheko-Reborn/nheko/commit/d94ac86816f9f325cba11f71344a3ca99591130d

I accidentally broke room joins in 0.7.x and released 1.7.1 just before I noticed that.

It will take a bit until the next release, that includes this fix.

**Comment by doronbehar** [comment]
> @doronbehar I can't just blindly merge any of these and I haven't had time yet to look at those in detail. I propose you use the workaround for now with a comment that it can be removed once #85254 and #73940 are resolved.

OK, done.

> Btw, you may want to include this commit in your package: Nheko-Reborn/nheko@d94ac86

Thanks @deepbluev7 for joining in and suggesting this, I appreciate your effort you are putting in Nheko.

Unfortunately, and I may need help with this @veprbl , I get this error:

```
applying patch /nix/store/cgmic2qicjfqzjgddg2f7a2pabmmspma-room-ids-escape-patch
can't find file to patch at input line 3
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- a/src/ChatPage.cpp
|+++ b/src/ChatPage.cpp
--------------------------
File to patch:
Skip this patch? [y]
```

I tried to play with the `stripLen` argument to `fetchpatch` but it didn't help.

**Comment by doronbehar** [comment]
My bad: I tried to patch mtxclient instead :/.

**Comment by doronbehar** [comment]
@GrahamcOfBorg build nheko

**Comment by doronbehar** [comment]
I'm testing the darwin build with:

@GrahamcOfBorg build nheko

**Comment by deepbluev7** [comment]
The build error you get on Darwin looks like your clang is too old to support the deduction guides for std::array. We are only testing clang-6 in our CI. (Not sure, how you guys handle compiler versions.)

**Comment by doronbehar** [comment]
OK @deepbluev7 that certainly should be possible to fix but this needs a real Darwin machine. I added a comment that links here.

Besides that all is ready.

**Comment by doronbehar** [comment]
@GrahamcOfBorg build nheko

**Comment by doronbehar** [comment]
(Last force push removing obsolete comments)

@GrahamcOfBorg build nheko

**Comment by bqv** [comment]
@deepbluev7 clang-6? That's old.. @doronbehar you can specialcase darwin to use pkgs.llvmPackages_6.stdenv and might have better luck

**Comment by doronbehar** [comment]
> @doronbehar you can specialcase darwin to use pkgs.llvmPackages_6.stdenv and might have better luck

@bqv That's not that trivial, because we use `mkDerivation` by `qt5`... It's too much effort for me to test this, as my only means are GrahamcOfBorg.

**Comment by rnhmjoj** [comment]
@doronbehar, sorry I don't know the details because I'm not a cmake user. It should disable the code responsible for finding boost. I think it's an acceptable workaround until boost is fixed.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 2]
```diff
@@ -1,70 +1,72 @@
...
+, tweeny
+, mkDerivation
+, qtbase
+, qtmacextras
+, qtmultimedia
+, qttools
+, qtquickcontrols2
+, qtgraphicaleffects
+, mtxclient
+, boost17x
```
Comment (rnhmjoj) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:18]: I think you should keep `boost` here and call the package with `boost = boost17x` in all-packages.nix.

**Review by rnhmjoj — COMMENTED** [review]
Just tested: builds and seems to work fine.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix (hunk 2 above)
Comment (doronbehar) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:18]: Some people like it this way, and some people like it as I did :/ whenever I do either I'm commented... I don't have a strong opinion on this repetitive dilemma but there's no convention that I'm aware of. Maybe @rnhmjoj you could spare me please another cycle of Git checkout -> edit -> ... Git commit -> Git push -f ?

Comment (rnhmjoj) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:18]: Uhm, you're right: I checked with openssl and boost (two major sources of incompatible versions) and it seems we are really split 50-50 on this topic. I think the version should be treaded like an override and kept out of the nix expression of a package but given there is no unanimity let's have it your way.

**Review by rnhmjoj — APPROVED** [review]

**Comment by deepbluev7** [comment]
@bqv clang-6 is just the minimum version, anything newer should work (or even apple clang in Xcode 10.2+). But the CI job used clang-5, which seems to be too old.

**Comment by rnhmjoj** [comment]
> @bqv That's not that trivial, because we use mkDerivation by qt5... It's too much effort for me to test this, as my only means are GrahamcOfBorg.

`libsForQt5.mkDerivation` is very simple. Basically you can replicate it just by adding `wrapQtAppsHook`. See pkgs/development/libraries/qt-5/mkDerivation.nix.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 3]
```diff
@@ -1,70 +1,72 @@
...
+    spdlog
+    cmark
+    qtbase
+    qtmultimedia
+    qttools
+    qtquickcontrols2
+    qtgraphicaleffects
   ] ++ lib.optional stdenv.isDarwin qtmacextras;
 
   enableParallelBuilding = true;
```
Comment (rnhmjoj) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:72]: I'm pretty sure this is redundant.

Comment (doronbehar) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:72]: Why?

**Comment by doronbehar** [comment]
@GrahamcOfBorg build

File: pkgs/applications/networking/instant-messengers/nheko/default.nix (hunk 3 above)
Comment (rnhmjoj) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:73]: It should be already `true`, at least for Qt applications.

Comment (doronbehar) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:73]: That's correct.

**Comment by doronbehar** [comment]
The Darwin build still fails, but differently:

https://logs.nix.ci/?key=nixos/nixpkgs.85922&attempt_id=1d08c688-e7d7-45c0-bb00-c25f49ce9651

**Comment by deepbluev7** [comment]
@doronbehar The current build fails, because some C++17 library features are only available in macOS 10.14.

**Comment by doronbehar** [comment]
> @doronbehar The current build fails, because some C++17 library features are only available in macOS 10.14.

Thank you @deepbluev7 for replying. If so, we are almost there but I'm afraid I couldn't find a way to mark it as broken only if the minimum macOS version is not met... Hence I'm marking it as broken for every macOS version for now. If anyone knows a way to do that, please tell me. For now, the PR should be ready.

File: pkgs/development/libraries/mtxclient/default.nix [hunk 4]
```diff
@@ -1,5 +1,17 @@
...
+, fetchpatch
+, cmake
+, pkgconfig
+, boost
+, openssl
+, zlib
+, libsodium
+, olm
+, spdlog
+, gtest
```
Comment (veprbl) [inline pkgs/development/libraries/mtxclient/default.nix:12]: Commit message says "mtxclient: format inputs and arguments", but you add unused `gtest` argument.

File: pkgs/development/libraries/mtxclient/default.nix [hunk 5]
```diff
@@ -18,8 +30,18 @@ stdenv.mkDerivation rec {
...
   ];
 
-  nativeBuildInputs = [ cmake pkgconfig ];
-  buildInputs = [ boost openssl zlib libsodium olm ];
+  nativeBuildInputs = [
+    cmake
+    pkgconfig
+  ];
+  buildInputs = [
+    spdlog
```
Comment (veprbl) [inline pkgs/development/libraries/mtxclient/default.nix:44]: Commit message says "mtxclient: format inputs and arguments", but you add `spdlog`.

File: pkgs/development/libraries/mtxclient/default.nix [hunk 6]
```diff
@@ -33,6 +35,7 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [
     cmake
     pkgconfig
+    gtest
```
Comment (veprbl) [inline pkgs/development/libraries/mtxclient/default.nix:41]: Goes to `checkInputs`. But no point if you don't enable tests.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 7]
```diff
@@ -1,6 +1,20 @@
-{ lib, stdenv, fetchFromGitHub
-, cmake, cmark, lmdb, mkDerivation, qtbase, qtmacextras
-, qtmultimedia, qttools, mtxclient, boost, spdlog, olm, pkgconfig
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, cmark
+, lmdb
+, lmdbxx
```
Comment (veprbl) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:8]: Commit message says "nheko: format inputs and arguments", but you add an unused `lmdbxx` argument.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 8]
```diff
@@ -76,9 +77,6 @@ mkDerivation rec {
     homepage = "https://github.com/Nheko-Reborn/nheko";
     maintainers = with maintainers; [ ekleog fpletz ];
     platforms = platforms.all;
-    # Should be fixable if a higher clang version is used, see:
-    # https://github.com/NixOS/nixpkgs/pull/85922#issuecomment-619287177
-    broken = stdenv.targetPlatform.isDarwin;
```
Comment (veprbl) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix]: Could squash with the previous darwin commit.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 9]
```diff
@@ -75,6 +75,9 @@ stdenv.mkDerivation rec {
     homepage = "https://github.com/Nheko-Reborn/nheko";
     maintainers = with maintainers; [ ekleog fpletz ];
     platforms = platforms.all;
+    # Should compile for MacOS 10.14 or higher but I don't know a way to mark
+    # it as broken only under this condition, fixes for this are welcome
+    broken = stdenv.isDarwin;
```
Comment (veprbl) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:80]: Could squash with the previous darwin commit.

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -13633,7 +13635,10 @@ in
 
   mtpfs = callPackage ../tools/filesystems/mtpfs { };
 
-  mtxclient = callPackage ../development/libraries/mtxclient { };
+  mtxclient = callPackage ../development/libraries/mtxclient {
+    # https://github.com/NixOS/nixpkgs/pull/85922#issuecomment-619287177
+    stdenv = if stdenv.isDarwin then llvmPackages_9.stdenv else stdenv;
```
Comment (veprbl) [inline pkgs/top-level/all-packages.nix:13640]: Given the comment above, this doesn't make sense, we are using clang-7 and upstream tests with clang-6. Anyway, if this doesn't help to fix the build, we probably don't need to take in this trivial change.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix [hunk 11]
```diff
@@ -7,7 +7,7 @@
 , lmdb
 , lmdbxx
 , tweeny
-, mkDerivation
+, wrapQtAppsHook
```
Comment (veprbl) [inline pkgs/applications/networking/instant-messengers/nheko/default.nix:10]: I wonder why you do that?

**Review by veprbl — APPROVED** [review]
Passes nixpkgs-review on NixOS, runs, diff looks mostly okay, could use some (light) squashing.

File: pkgs/applications/networking/instant-messengers/nheko/default.nix (hunk 11 above)
Comment (doronbehar) [reply pkgs/applications/networking/instant-messengers/nheko/default.nix:10]: There's no need for that,I was confused between the effect of `libsForQt5.callPackage` and the fact `qt5` has it's own `mkDerivation`.

File: pkgs/top-level/all-packages.nix (hunk 10 above)
Comment (doronbehar) [reply pkgs/top-level/all-packages.nix:13640]: I got the impression that using this stdenv does help, but I'll put my mind to it no more. I'm putting a single commit for both mtxclient and nheko with a link to that comment by upstream.

**Review by doronbehar — COMMENTED** [review]
Thanks @veprbl for catching all of these imperfections. I went through all the commits and made sure they all line up to your comments.

**Pull request merged**